openpdf/src/main/java/com/lowagie/text/pdf/PdfStamperImp.java

```diff
@@ -870,17 +884,104 @@ void flatFields() {
                 if (page == -1)
                 	continue;
                 PdfDictionary appDic = merged.getAsDict(PdfName.AP);
+                PdfStream appStream=null;
+                
+                if (appDic != null) {
+                    appStream = appDic.getAsStream(PdfName.N);
+                }
+
+                //Lonzak (fix) if NeedAppearances flag is true then regenerate appearance before flattening
+                if (needAppearance!=null && needAppearance.booleanValue()) {
+                    
+                	boolean regenerate = false;
+
+                	//not existing AP
+                	if((appDic == null || appDic.get(PdfName.N) == null)) {
+                		regenerate=true;
+                	}
+                	else if(appDic.getDirectObject(PdfName.N) instanceof PdfIndirectReference) {
+                		//since newly added
+                		regenerate=false;
+                	}
+                	else {
+	                	int type = acroFields.getFieldType(name);
+	                    String value = acroFields.getField(name);
+	                                    		
+	                    //workaround for libre/open office which creates nearly empty streams: /TX BMC\nEMC
+	                    //Currently only for Textfields - for Radios/Checkboxes the appearance stream has to be determined (by looking at /AS or /V)
+	                    if(type==AcroFields.FIELD_TYPE_TEXT && appStream instanceof PRStream) {
+		                    if(value!=null && !value.isEmpty()) {
+		                    	try {
+		        					byte[] bytes= PdfReader.getStreamBytes((PRStream)appStream);
+		        					((PRStream)appStream).setData(bytes);
+		        					if(new String(bytes).equals("/Tx BMC\nEMC\n")) {
+		        						regenerate=true;
+		        					}
+		        				}
+		        				catch (IOException e) {
+		        					//ignore
+		        				}
+		                    }
+	                    }
+                	}
+                	
+                	if(regenerate) {
+                		try {
+	                        this.acroFields.regenerateField(name);
+	                        appDic = this.acroFields.getFieldItem(name).getMerged(k).getAsDict(PdfName.AP);
+	                    }
+	                    catch (Exception e) {
+	                    	//ignore any exception
+	                    }
+                	}
+                }
+                
+                boolean transformNeeded=false;
+                double rotation = 0;
+                if(merged.getAsDict(PdfName.MK) != null && merged.getAsDict(PdfName.MK).get(PdfName.R) != null){
+                    rotation = merged.getAsDict(PdfName.MK).getAsNumber(PdfName.R).floatValue();
```

Comment:
Ah ok interesting - so should I leave the rotation code in there in case there is no matrix defined?


                double rotation = 0;
                if(merged.getAsDict(PdfName.MK) != null && merged.getAsDict(PdfName.MK).get(PdfName.R) != null){
                    rotation = merged.getAsDict(PdfName.MK).getAsNumber(PdfName.R).floatValue();
                }
                if (appDic != null && appDic.getAsArray(PdfName.MATRIX) != null) {

                    PdfArray matrixArray = appDic.getAsArray(PdfName.MATRIX);
                    
                    if(matrixArray.size() == 6) {
                        float a = matrixArray.getAsNumber(0).floatValue();
                        float b = matrixArray.getAsNumber(1).floatValue();

                        double rotationRadians = Math.atan2(b, a);
                        //in degrees
                        rotation = Math.toDegrees(rotationRadians);
                    }
                }

Reply:
Working with the matrix should be done strictly according to spec, i.e. using the algorithm in 12.5.5. If that looks to complicated, continue using the *MK*/*R* value

Reply:
The spec says something like this:

Algorithm: appearance streams

1.  The appearance’s bounding box (specified by its BBox entry) shall be transformed, using Matrix, to produce a quadrilateral with arbitrary orientation. The transformed appearance box is the smallest upright rectangle that encompasses this quadrilateral.
2.  A matrix A shall be computed that scales and translates the transformed appearance box to align with the edges of the annotation’s rectangle (specified by the Rect entry). A maps the lower-left corner (the corner with the smallest x and y coordinates) and the upper-right corner (the corner with the greatest x and y coordinates) of the transformed appearance box to the corresponding corners of the annotation’s rectangle.
3.  Matrix shall be concatenated with A to form a matrix AA that maps from the appearance’s coordinate system to the annotation’s rectangle in default user space:
AA = Matrix × 𝐴

Ok I think we'll stick with /R for now....